www/src/pages/semesters/[semesterSlug]/api/subject-copy.ts

```diff
@@ -0,0 +1,91 @@
+import type { APIRoute } from 'astro';
+import { Classroom } from '@backend/classroom';
+import { Exercise, type ExerciseCreateData } from '@backend/exercise';
+import { LaboratoryGroup, type LaboratoryGroupCreateData } from '@backend/laboratory-group';
+import { Semester } from '@backend/semester';
+import { Subject } from '@backend/subject';
+import { subjectCopyFromPreviousSemesterApiSchema } from '@components/subjects/types';
+
+export const POST: APIRoute = async ({ locals, params }) => {
+    const { jsonData, user } = locals;
+
+    if (!user) {
+        return Response.json(null, { status: 404 });
+    }
+
+    const currentSemester = await Semester.fetchBySlug(params.semesterSlug ?? '');
+
+    if (!currentSemester) {
+        return Response.json(null, { status: 404 });
+    }
+
+    const data = subjectCopyFromPreviousSemesterApiSchema.nullable().catch(null).parse(jsonData);
+
+    if (!data) {
+        return Response.json(null, { status: 400 });
+    }
+
+    const semesterToCopyFrom = await Semester.fetch(data.semesterId);
+
+    if (!semesterToCopyFrom) {
+        return Response.json(null, { status: 404 });
+    }
+
+    const subjectToBeCopied = await Subject.fetch(data.subjectId);
+
+    if (!subjectToBeCopied) {
+        return Response.json(null, { status: 404 });
+    }
+
+    const currentSemesterSubjects = await Subject.fetchAllFromSemester(currentSemester);
+
+    // Copy subject
+    if (currentSemesterSubjects.find(s => s.name.toLowerCase() === subjectToBeCopied.name.toLowerCase())) {
+        return Response.json(false, { status: 200 });
+    }
+
+    const newSubject = await Subject.create({
+        name: subjectToBeCopied.name,
+        semester: currentSemester,
+        teachers: await subjectToBeCopied.getTeachers(),
+        description: subjectToBeCopied.description,
+        moodleCourseId: '', // Course id is not copied
+    });
+
+    // Copy exercises
+    const exercisesToBeCopied = await Exercise.fetchAllFromSubject(subjectToBeCopied);
+    const exercisesCreateData: ExerciseCreateData[] = [];
+    for (const exercise of exercisesToBeCopied) {
+        const classroom = await Classroom.fetch(exercise.classroomId);
+
+        if (!classroom) {
+            return Response.json(null, { status: 400 });
+        }
```

Comment:
After merging #113 this should be removed to allow copying a subjects containing exercise with null classroom.